[iOS] imported/w3c/web-platform-tests/fetch/security/dangling-markup-mitigation-data-url.tentative.sub.html is a flaky failure

youennf opened the pull request:
#### 76bace0870c361333cc42ae21e89f26cf6f37d2e
<pre>
[iOS] imported/w3c/web-platform-tests/fetch/security/dangling-markup-mitigation-data-url.tentative.sub.html is a flaky failure
<a href="https://rdar.apple.com/165165272">rdar://165165272</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=302901">https://bugs.webkit.org/show_bug.cgi?id=302901</a>

Reviewed by Alex Christensen.

The load event is fired too soon for documents that reuse a loading subresource loader via the memory cache.
Previously, since the load would not trigger a new SubresourceLoader, the existing mechanism to wait for loads to finish before firing the load event would be broken.

We use the current request count mechanism in CachedResourceLoader as follows:
- When a resource is reused for a load and the resource is loading, we increment the request count.
- CachedResourceLoader registers itself to know when the resource is no longer loading. When this happens, we decrement the request count and check whether the load event should fire.

We introduce in CachedResource a callback mechanism to know when a resource is no longer loading.
Covered by test being no longer flaky.

Canonical link: <a href="https://commits.webkit.org/304412@main">https://commits.webkit.org/304412@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1460586a96a4c645fdd82abf4189b94f08cd81f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135358 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46632 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143049 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/87079 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9b7f4a6f-19a7-4b71-92a5-12f18b89992c) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137227 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8375 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7586 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103447 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e607554b-90eb-4e18-a3ca-5d5efe154e43) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138304 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5991 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121322 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84312 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/caa47e76-d21d-4eb7-b988-d694b37baedb) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5775 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3382 "Found 1 new API test failure: TestWebKitAPI.WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginAndEndEditingEvents (failure)") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3659 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114979 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39506 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145804 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7422 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40076 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111825 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7464 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6219 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112194 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28489 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5628 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117622 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61335 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7476 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35748 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7224 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71023 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7445 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7327 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/loader/cache/CachedResource.cpp:
```diff
@@ -376,6 +380,22 @@ void CachedResource::finish()
+        callback();
+}
+
+void CachedResource::whenLoaded(Function<void()>&& callback)
```
It seems like this is called once and exactly once.  It could be a CompletionHandler.

youennf replied:
Done

youennf committed:
[iOS] imported/w3c/web-platform-tests/fetch/security/dangling-markup-mitigation-data-url.tentative.sub.html is a flaky failure
rdar://165165272
https://bugs.webkit.org/show_bug.cgi?id=302901

Reviewed by Alex Christensen.

The load event is fired too soon for documents that reuse a loading subresource loader via the memory cache.
Previously, since the load would not trigger a new SubresourceLoader, the existing mechanism to wait for loads to finish before firing the load event would be broken.

We use the current request count mechanism in CachedResourceLoader as follows:
- When a resource is reused for a load and the resource is loading, we increment the request count.
- CachedResourceLoader registers itself to know when the resource is no longer loading. When this happens, we decrement the request count and check whether the load event should fire.

We introduce in CachedResource a callback mechanism to know when a resource is no longer loading.
Covered by test being no longer flaky.

Canonical link: https://commits.webkit.org/304412@main
